state.go

```diff
@@ -922,16 +959,19 @@ func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (Sta
 
 	log.Debug("handle update status report state")
 
-	if err := StoreStateData(ctx.store, StateData{
-		Name:         usr.Id(),
-		UpdateInfo:   usr.Update(),
-		UpdateStatus: usr.status,
-	}); err != nil {
-		log.Errorf("failed to store state data in update status report state: %v",
-			err)
-		return NewReportErrorState(usr.Update(), usr.status), false
+	// Do not store this if artifact-commit scripts are run when leaving the state
+	// as then the scripts will not be rerun
+	if usr.Transition() != ToArtifactCommit {
```

Comment:
So why we are storing the state now only if we are not entering here from commit state? 
Next, the fact that we need to use `Transition()` here means we need to really think about refactoring this whole state machine concept as it is becoming quite messy.